---
title: "TCR Clonotype Analysis of CITEseq dataset"
author: "Gillian Savage"
date: "1/9/2025"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(Seurat) #using version 4.4
library(tidyverse)
library(RColorBrewer)
library(ggpubr)
library(ggplot2)
library(immunarch)
library(scRepertoire)

mutate <- dplyr::mutate
filter <- dplyr::filter
count <- dplyr::count
summarise <- dplyr::summarise
rename <- dplyr::rename
select <- dplyr::select
options(dplyr.summarise.inform=FALSE)

```

# Load Data

Load in processed and mapped Seurat Object of 5'GEX data
```{r read data}
#read in mapped 5'GEX seurat object
sobjs_T_5prime <- readRDS("~/citeseq_lymphoma/output/SeuratObjects_T_5prime.rds")
```


Load in annotated TCR contig data. This was generated by running `cellranger vdj` on 5'TCR fastq files obtained from the European Genome Archive ([EGAD50000000497](https://ega-archive.org/datasets/EGAD50000000497)).
```{r}
#load contig data

LN0144 <- read.csv("~/trm_dlbcl/external_data/contigs/FL3/filtered_contig_annotations.csv")
LN0278 <- read.csv("~/trm_dlbcl/external_data/contigs/FL9/filtered_contig_annotations.csv")
LN0198 <- read.csv("~/trm_dlbcl/external_data/contigs/FL11/filtered_contig_annotations.csv")
LN0178 <- read.csv("~/trm_dlbcl/external_data/contigs/GCB2/filtered_contig_annotations.csv")
LN0193 <- read.csv("~/trm_dlbcl/external_data/contigs/GCB3/filtered_contig_annotations.csv")
LN0110 <- read.csv("~/trm_dlbcl/external_data/contigs/MCL3/filtered_contig_annotations.csv")
LN0078 <- read.csv("~/trm_dlbcl/external_data/contigs/MCL4/filtered_contig_annotations.csv")
LN0302 <- read.csv("~/trm_dlbcl/external_data/contigs/MZL9/filtered_contig_annotations.csv")
LN0217 <- read.csv("~/trm_dlbcl/external_data/contigs/MZL11/filtered_contig_annotations.csv")
LN0259 <- read.csv("~/trm_dlbcl/external_data/contigs/rLN1/filtered_contig_annotations.csv")
LN0132 <- read.csv("~/trm_dlbcl/external_data/contigs//rLN5/filtered_contig_annotations.csv")

#make contig list
contig_list <- list(LN0144, LN0278, LN0198, LN0178, LN0193, LN0110, LN0078, LN0302, LN0217, LN0259, LN0132)
```

# Analysis of TCR Clonality per cluster
Combine all TCR data
```{r}
combined.TCR <- combineTCR(contig_list,
                           removeNA = FALSE, 
                           removeMulti = FALSE, 
                           filterMulti = FALSE)

#add corresponding sample ID
combined.TCR <- addVariable(combined.TCR, 
                            variable.name = "sample", 
                            variables =  c("LN0144", "LN0278", "LN0198", "LN0178", "LN0193", "LN0110", "LN0078", "LN0302", "LN0217", "LN0259", "LN0132"))

# view output
head(combined.TCR[[1]])
```

Combine with 5'GEX data to have cluster location for each TCR sequence and analyze clonality per cluster (Supplementary Fig. S8A)
```{r}

#change cluster value to factor
sobjs_T_5prime$predicted.celltype <- as.factor(sobjs_T_5prime$predicted.celltype)

#combine TCR data and 5'GEX Seurat object
Tcells <- combineExpression(combined.TCR, 
                                   sobjs_T_5prime, 
                                   cloneCall="CTgene",
                                   proportion = FALSE,
                            group.by = "sample",
                                   cloneSize=c(Single=1, Small=5, Medium=20, Large=100, Hyperexpanded=500))


#visualize clonality per patient and cluster
clonalOccupy(Tcells, 
             x.axis = "subject_id", facet.by = "predicted.celltype", 
             proportion = T, 
             label = FALSE,
             palette = "Spectral")


```

# Analysis of viral TCR clonotypes

Use Immunarch from combined airr file generated by running `cellranger vdj` on 5'TCR fastq files obtained from the European Genome Archive ([EGAD50000000497](https://ega-archive.org/datasets/EGAD50000000497)).
```{r}

file_path = "~/citeseq_lymphoma/vdj_output/contigs/"

#load all files
immdata <- repLoad(file_path)
```

```{r}
scdata_cl <- select_clusters(immdata, sobjs_T_5prime$predicted.celltype, "Cluster")
scdata_cl$meta
```

Annotate with Viral Epitopes
```{r}
#load database of viral TCR sequences obtained from https://vdjdb.cdr3.net/

VDJdb <- read_tsv("~/trm_dlbcl/external_data/VDJdb.tsv")

#annotate TCR sequences if they are viral using immunarch
annotated <- dbAnnotate(scdata_cl$data, VDJdb, "CDR3.aa", "CDR3")

# select for CDR3 sequences and species of the epitope and remove duplicate CDR3 sequences so epitope data can be added to annotated TCR data table
VDJdb.ep <- select(VDJdb, c("CDR3", "Epitope species"))
VDJdb.ep <- VDJdb.ep[!duplicated(VDJdb.ep$CDR3), ]
 
#add epitope info for CDR3 sequences
annotated.m <- merge(annotated.m, VDJdb.ep, by.x = c("CDR3.aa"), by.y = c("CDR3"), all.x = F)

#output csv file of viral clones per sample and cluster
write_csv(annotated.m, "~/trm_dlbcl/external_data/annotated_virus.csv")
```

Get average number of clones per cluster to be able to calculate the proportion of clones that are viral in each cluster
```{r}
clones <- repExplore(scdata_cl$data, .method = "clones", .col = "aa")

#output csv of clones per cluster
write_csv(clones, "~/trm_dlbcl/external_data/clones_per_clus.csv")
```

Outputted csv of viral clones per patient and cluster and the number of clones per cluster and patient were used to calculate the proportion of clones per cluster that were virus-specific in each patient. This data was then graphed in GraphPad Prism and displayed in Supplementary Fig. S8B.